### DISCUSSION TITLE
How to go about wrapping GtkBuilder?

### DISCUSSION THREAD

**Discussion by PhilippMDoerner** [can-lehmann/owlkettle#138] [Q&A] [answered]
I stumbled over this as part of trying to wrap the ShortcutWindow widget.
It appears there's no way to create it without a [GtkBuilder](https://docs.gtk.org/gtk4/class.Builder.html).

Given that, we'll need to start wrapping that API.
How it generally works is that you define essentially XML as a "ui" file, load that into a builder and then... somehow it gets shown.

I've begun the works on that, but where I'm struggling is once I have a GtkBuilder instance with the necessary UI-stuff loaded, to create a GtkWidget from that.

I've stumbled over [this](https://github.com/ToshioCP/Gtk4-tutorial/blob/main/gfm/sec9.md#gtkbuilder) example that provides some code on how to do this, but I have some syntax problem there.
The steps appear to be:

1. Create Builder with UI file
2. Fetch pointer to GObject from Builder instance using `gtk_builder_get_object` and the id of the tag defined in the UI file
3. Use the C Syntax `GTK_WIDGET (thatGObjectPointer)` to create a GTKWidget instance

I don't have nim syntax to replicate 3, so I can't create a GTKWidget. Are there any ways to copy what they do somehow?

Edit: I tried in the GTK Matrix channel for some advice and was directed to the [GObject](https://docs.gtk.org/gobject/concepts.html#instantiatable-classed-types-objects) docs. While I think they're useful if you already have some C experience, they're rather complex when you don't and so trying to wrap my head around what was written wasn't all that successful :sweat_smile:

**Comment by PhilippMDoerner** [comment, marked as answer]
This took way too long but I *finally* managed to get a working example going:
```nim
#builder.nim
import ./bindings/gtk

type 
  BuilderObj = object
    gtk*: GtkBuilder

  Builder* = ref BuilderObj

proc wrap(builder: GtkBuilder): Builder = Builder(gtk: builder)
  
proc newBuilderFromFile*(filename: string): Builder =
  let builder = gtk_builder_new_from_file(filename.cstring)
  result = builder.wrap()

proc newBuilderFromResource*(resource: string): Builder =
  let builder = gtk_builder_new_from_resource(resource.cstring)
  result = builder.wrap()

proc newBuilderFromString*(uiString: string): Builder =
  let builder = gtk_builder_new_from_string(uiString.cstring, uiString.len().cint)
  result = builder.wrap()

proc newBuilder*(): Builder = gtk_builder_new().wrap()

proc getWidget*(builder: Builder, id: string): GtkWidget =
  let gObj: pointer = gtk_builder_get_object(builder.gtk, id.cstring)  
  return gObj.GtkWidget

proc newWidgetFromString*(uiString: string, id: string = "widget"): GtkWidget =
  let builder = newBuilderFromString(uiString)
  return builder.getWidget(id)

proc newWidgetFromFile*(fileName: string, id: string = "widget"): GtkWidget =
  let builder = newBuilderFromFile(fileName)
  return builder.getWidget(id)

proc newWidgetFromResource*(resource: string, id: string = "widget"): GtkWidget =
  let builder = newBuilderFromResource(resource)
  return builder.getWidget(id)
```


```nim
#example.nim
import std/[sequtils]
import owlkettle, owlkettle/[dataentries, builder, playground, adw]
import owlkettle/bindings/gtk

const uiString = """
  <?xml version="1.0" encoding="UTF-8"?>
  <interface>
    <object class="GtkButton" id="widget">
      <property name="label">Hello, world</property>
    </object>
  </interface>
"""

renderable BuilderExample1:
  hooks:
    beforeBuild:
      state.internalWidget = newWidgetFromString(uiString, "widget")


viewable App:
method view(app: AppState): Widget =
  result = gui:
    Window():
      Box(orient = OrientY):
        BuilderExample1()

adw.brew(gui(App()))
```

**Comment by can-lehmann** [comment]
`GTK_WIDGET` just maps to `GtkWidget(...)`, but judging by your forum post, you already figured that out. There is actually no need for `Builder` here, even if a widget does not have a constructor, it can still be instantiated: https://github.com/can-lehmann/owlkettle/blob/main/owlkettle/widgets.nim#L2107

**Reply by PhilippMDoerner** [reply]
That... would've saved me 5h to know beforehand :sweat_smile: 

Hmm... in that case I guess the effort in builder is kinda pointless given we typically don't use it and appear to have no usecase?
I mean I could imagine it as being "nice" for users to define their own renderables but I don't see any direct usecase tbh.

**Reply by can-lehmann** [reply]
Sorry, I only saw your post just now.
The only usecase I can think of for builder right now is if sb. has an existing app they want to port to owlkettle. However that seems like a very niche use case.

**Reply by PhilippMDoerner** [reply]
Nah, all good, got to learn step by step.

In that case could we still introduce it **solely** for ShortcutsWindow?
I am stating this because [ShortCutsWindow](https://docs.gtk.org/gtk4/class.ShortcutsWindow.html) has no proper method (yet) to add Sections.
[Sections](https://docs.gtk.org/gtk4/class.ShortcutsSection.html) don't have a proper method (yet) either to add [Groups](https://docs.gtk.org/gtk4/class.ShortcutsGroup.html), which in turn don't have proper methods to add  Shortcuts to themselves.

The docs there state that building these via builder is the preferred way. And until at least the next minor GTK release (4.14) I don't see any non-builder way to do this Widget.

**Reply by can-lehmann** [reply]
Hmm, thats indeed an issue. You can use the `Buildable` interface to add children (this is what Builder uses internally): https://docs.gtk.org/gtk4/vfunc.Buildable.add_child.html. However it does not provide a way to remove children again afaik.

**Reply by PhilippMDoerner** [reply]
I think particularly for ShortcutsWindow it'd be fair to just have that window be static - Shortcuts are shortcuts, if those change then only at compile-time of the next version of the application.
In which case *either* I try to use Buildable.add_child *or* we add the builder functionality and generate the ShortcutsWindow from there.
Either by generating the GtkWindow directly in some proc or I figure out whether we can have a renderable ShortcutsWindow that makes use of the builder and accepts a uiString to construct said window around.

Sound good?

**Reply by PhilippMDoerner** [reply]
Answer please :stuck_out_tongue:

**Reply by can-lehmann** [reply]
I would prefer using `Buildable.add_child` over `Builder`, as constructing an XML UI description stands in direct contrast to the aims of owlkettle. If there is truely no way to make it dynamic, thats fine for now.
